### PR TITLE
Update Server and Direct gateways to support protocol version 4.0

### PR DESCRIPTION
**Pull request by nfourtythree**
This pull request provides an update to the protocol version for the impending 14th March deadline imposed by Opayo/SagePay. The PR is only related to the change in protocol and **does not** add any other integrations.

Both the Direct and Server gateways are updated to work with the new protocol.

**Points of note**
- ~Reverted PHP version requirement back to `7.2` as our Craft CMS product has `7.2.5` as the lowest version required.~ Left the version as `^7.3` due to dependency requirements.
- Hardcoded the change to the protocol version to `4.00`, this means there is a need for the next release (when merged) to be a major version to avoid breaking anyone's projects.

Merging this will help us release a new version of our [SagePay plugin](https://github.com/craftcms/commerce-sagepay) to meet the upcoming requirement changes being enforced by Opayo.

This will also resolve the craftcms/commerce-sagepay#20 issue that we have outstanding, in which we will create a new release for people to update to after this package has been updated.

Let me know your feedback and thoughts, thanks!

### CODE REVIEW

**Comment by barryvdh** [comment]
Thanks, can you look into the tests?

File: src/Message/DirectAuthorizeRequest.php [hunk 1]
```diff
@@ -56,6 +56,30 @@ protected function getBaseAuthorizeData()
         $data['VendorTxCode'] = $this->getTransactionId();
         $data['ClientIPAddress'] = $this->getClientIp();
 
+        $data['BrowserJavascriptEnabled'] = $this->getBrowserJavascriptEnabled() ?: static::BROWSER_JAVASCRIPT_NO;
+        $data['BrowserLanguage'] = $this->getBrowserLanguage() ?: static::BROWSER_LANGUAGE;
+        $data['ThreeDSNotificationURL'] = $this->getThreeDSNotificationURL();
+        $data['BrowserAcceptHeader'] = $_SERVER['HTTP_ACCEPT'];
```
Comment (barryvdh) [inline src/Message/DirectAuthorizeRequest.php:62]: I would just add  `?? null`

File: src/Message/DirectAuthorizeRequest.php [hunk 2]
```diff
@@ -56,6 +56,30 @@ protected function getBaseAuthorizeData()
...
+        $data['BrowserUserAgent'] = $_SERVER['HTTP_USER_AGENT'];
```
Comment (barryvdh) [inline src/Message/DirectAuthorizeRequest.php:63]: Same here, just add `?? null`

File: src/Message/DirectAuthorizeRequest.php (hunk 1 above)
Comment (nfourtythree) [reply src/Message/DirectAuthorizeRequest.php]: Have done this, errors relating to these lines weren't showing when running the tests locally.

**Comment by nfourtythree** [comment]
Just to note I have also reverted the PHP change. There are too many dependencies issues to go back down to `^7.2` so have bumped it back to `^7.3`

**Comment by barryvdh** [comment]
Branch alias is bumped to 4.0 so you can test with `^4@dev`

**Comment by jamieburchell** [comment]
@nfourtythree @barryvdh What happens when this is introduced in to a project where there are pending protocol v3 transactions? Will the signature checks fail/mismatch, or is it OK because the new fields will be empty strings and no affect the signature?

**Comment by steveWinter** [comment]
> Branch alias is bumped to 4.0 so you can test with `^4@dev`

Unfortunately this doesn't seem to work?
```Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires league/omnipay ^4@dev,  found league/omnipay[dev-feat/add-tests, dev-master, dev-guzzle6-adapter-v2, dev-judgej-patch-1, dev-delatbabel-patch-1, dev-feat-php8, dev-barryvdh-travis-73_74, v0.6.0, ..., v0.9.3, v1.0.0, ..., 1.1.x-dev, v2.0.0, ..., 2.3.x-dev, v3.0-alpha.1, ..., 3.2.x-dev (alias of dev-master)] but it does not match the constraint.```
 
Did I misunderstand?

**Comment by barryvdh** [comment]
No clue

**Comment by steveWinter** [comment]
> No clue

Apparently neither had I (hence why I deleted the comment) - I'd been trying to update `league/omnipay` and not `omnipay/sagepay` 🤦

**Comment by barryvdh** [comment]
Yeah no I saw your mistake there but your comment was gone 😅

No clue was about the switch between V3 and V4 mid transaction

**Comment by jamieburchell** [comment]
I tried it - registered the transaction with old protocol 3 code, upgraded to protocol 4 code, completed transaction. It worked and showed protocol version 4 in the notification data. I was expecting it to say 3, since the documentation says it's whatever was supplied originally, but whatever.

**Comment by terryupton** [comment]
I left a note on the Craft Commerce thread <https://github.com/craftcms/commerce-sagepay/issues/20> 

We have tested this on both a standard install and a slightly custom install with some additional functionality around tokens and both sites are working fine and transactions are working as expected.

**Pull request merged**